### PR TITLE
[WIP] account softirq energy separately

### LINKED ISSUE #252: Attributing block I/O energy
**Is your feature request related to a problem? Please describe.**
Currently the attribution to Pods is based the Pods CPU/memory activities. The processes in the Pods could have disk I/O activities that are handled by interrupt handlers, we should attribute energy consumed by the handlers properly to the Pods that initiate them.

**Describe the solution you'd like**
The I/O stats are collected by cgroupfs. The interrupt handlers are currently classified as system processes. 
In the first implementation, we can do the following flow
- get the kworker process(es) that handle(s) the disk I/Os, quantify the processes' energy consumption.
- from the process's cgroup disk I/O stat, find the block I/O stats on each block device
- evenly distribute the kworker's energy to all processes based on their I/O stats
- create another metrics for the pods called Pod disk I/O energy

We can quantify other I/O energy in similar way, e.g. network I/O.

### PR DESCRIPTION
**Pull request by rootfs**
to fix #252 #256

### CODE REVIEW

File: pkg/collector/blockio_energy.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+package collector
+
+import (
+	"fmt"
+	"os"
+	"regexp"
+)
+
+var (
+	kblockdPodName       = "kworker-kblockd"
```
Comment (jichenjc) [inline pkg/collector/blockio_energy.go:26]: looks like this a process instead of independent pod?

Comment (rootfs) [reply pkg/collector/blockio_energy.go:26]: it is a name i want to give to all the kblockd worker threads, but you are right, the name shouldn't be exported and confuse end users.

Comment (marceloamaral) [reply pkg/collector/blockio_energy.go:26]: We could have a prefix for OS related process `os-process-kworker-kblockd`

**Comment by rootfs** [comment]
This is tested in the following way:

- curl the metrics endpoint and get kblockd container:
- run a `dd` program in the background `dd if=/dev/zero of=/home/test1g bs=1M count=1000`

`kblockd` energy metrics. The energy increases when `dd` is running and stablizes after `dd`
```console
# while true; do curl localhost:8888/metrics 2>&1 |grep kepler_container_package_joules_total |grep block; sleep 5;done
kepler_container_package_joules_total{command="",container_name="kblockd",container_namespace="system",pod_name="system_processes"} 0.124
kepler_container_package_joules_total{command="",container_name="kblockd",container_namespace="system",pod_name="system_processes"} 0.188
kepler_container_package_joules_total{command="",container_name="kblockd",container_namespace="system",pod_name="system_processes"} 0.199
kepler_container_package_joules_total{command="",container_name="kblockd",container_namespace="system",pod_name="system_processes"} 4.685
kepler_container_package_joules_total{command="",container_name="kblockd",container_namespace="system",pod_name="system_processes"} 12.365
kepler_container_package_joules_total{command="",container_name="kblockd",container_namespace="system",pod_name="system_processes"} 15.182
kepler_container_package_joules_total{command="",container_name="kblockd",container_namespace="system",pod_name="system_processes"} 17.27
kepler_container_package_joules_total{command="",container_name="kblockd",container_namespace="system",pod_name="system_processes"} 25.227
kepler_container_package_joules_total{command="",container_name="kblockd",container_namespace="system",pod_name="system_processes"} 28.753
kepler_container_package_joules_total{command="",container_name="kblockd",container_namespace="system",pod_name="system_processes"} 28.81
kepler_container_package_joules_total{command="",container_name="kblockd",container_namespace="system",pod_name="system_processes"} 28.847
kepler_container_package_joules_total{command="",container_name="kblockd",container_namespace="system",pod_name="system_processes"} 28.854
kepler_container_package_joules_total{command="",container_name="kblockd",container_namespace="system",pod_name="system_processes"} 29.063
kepler_container_package_joules_total{command="",container_name="kblockd",container_namespace="system",pod_name="system_processes"} 29.101
```

File: pkg/cgroup/kblockd.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+var (
+	processNameRegex     = "^kworker/"
+	processFullNameRegex = "^kworker/.*kblockd"
+	kblockCache          = make(map[uint64]bool)
+)
+
+func isKblockdWorker(comm string, pid uint64) bool {
+	found, _ := regexp.MatchString(processNameRegex, comm)
+	// if it is kworker, check the /proc/pid/comm to get full name
```
Comment (marceloamaral) [inline pkg/cgroup/kblockd.go:35]: Is it really necessary?
Reading files for all kernel processes is expensive.

**Comment by marceloamaral** [comment]
Looks good, I have only one comment.

**Comment by rootfs** [comment]
this is the new metrics
```console
kepler_container_package_joules_total{command="",container_name="NET_TX",container_namespace="system",pod_name="system_processes"} 0.001
kepler_container_package_joules_total{command="",container_name="BLOCK",container_namespace="system",pod_name="system_processes"} 0.065
kepler_container_package_joules_total{command="",container_name="system_processes",container_namespace="system",pod_name="system_processes"} 3.738
kepler_container_package_joules_total{command="",container_name="RCU",container_namespace="system",pod_name="system_processes"} 27.025
kepler_container_package_joules_total{command="",container_name="NET_RX",container_namespace="system",pod_name="system_processes"} 51.487
kepler_container_package_joules_total{command="",container_name="SCHED",container_namespace="system",pod_name="system_processes"} 280.341
```

**Comment by rootfs** [comment]
@marceloamaral updated, PTAL, thanks!

File: bpfassets/perf_event/perf_event.c [hunk 3]
```diff
@@ -224,9 +225,33 @@ int kprobe__finish_task_switch(struct pt_regs *ctx, struct task_struct *prev)
...
+    unsigned char common_flags; 
+    unsigned char common_preempt_count;
+    int common_pid;
+    unsigned int vec;
+} irq_entry_args_t;
+
+// per https://www.kernel.org/doc/html/latest/core-api/tracepoint.html#c.trace_softirq_entry
+int softirq_entry(irq_entry_args_t *args)
+{
+    u64 cur_pid = bpf_get_current_pid_tgid() >> 32;
```
Comment (rootfs) [inline bpfassets/perf_event/perf_event.c:248]: turns out this `cur_pid` is not softirq process id

**Comment by rootfs** [comment]
this PR doesn't seem to go anywhere to attribute time spent in block device or network to containers, closing now.

**Pull request closed without merging**